// x-pack/plugins/beats_management/public/containers/tags.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const unstated_1 = require("unstated");
class TagsContainer extends unstated_1.Container {
    constructor(libs) {
        super();
        this.libs = libs;
        this.reload = async (kuery) => {
            if (kuery) {
                this.query = await this.libs.elasticsearch.convertKueryToEsQuery(kuery);
            }
            else {
                this.query = undefined;
            }
            const tags = await this.libs.tags.getAll(this.query);
            this.setState({
                list: tags,
            });
        };
        this.delete = async (tags) => {
            const tagIds = tags.map((tag) => tag.id);
            const success = await this.libs.tags.delete(tagIds);
            if (success) {
                this.reload(this.query);
            }
            return success;
        };
        this.upsertTag = async (tag) => {
            const beatTag = await this.libs.tags.upsertTag(tag);
            if (beatTag !== null) {
                this.reload(this.query);
            }
            return beatTag !== null;
        };
        this.state = {
            list: [],
        };
    }
}
exports.TagsContainer = TagsContainer;
