// x-pack/plugins/upgrade_assistant/public/components/tabs.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const lodash_1 = require("lodash");
const react_1 = tslib_1.__importDefault(require("react"));
const eui_1 = require("@elastic/eui");
const react_2 = require("@kbn/i18n/react");
const chrome_1 = tslib_1.__importDefault(require("ui/chrome"));
const latest_minor_banner_1 = require("./latest_minor_banner");
const checkup_1 = require("./tabs/checkup");
const overview_1 = require("./tabs/overview");
const types_1 = require("./types");
class UpgradeAssistantTabsUI extends react_1.default.Component {
    constructor(props) {
        super(props);
        this.onTabClick = (selectedTab) => {
            const selectedTabIndex = lodash_1.findIndex(this.tabs, { id: selectedTab.id });
            if (selectedTabIndex === -1) {
                throw new Error(`Clicked tab did not exist in tabs array`);
            }
            this.setSelectedTabIndex(selectedTabIndex);
        };
        this.setSelectedTabIndex = (selectedTabIndex) => {
            this.setState({ selectedTabIndex });
        };
        this.loadData = async () => {
            try {
                this.setState({ loadingState: types_1.LoadingState.Loading });
                const resp = await axios_1.default.get(chrome_1.default.addBasePath('/api/upgrade_assistant/status'));
                this.setState({
                    loadingState: types_1.LoadingState.Success,
                    checkupData: resp.data,
                });
            }
            catch (e) {
                this.setState({ loadingState: types_1.LoadingState.Error });
            }
        };
        this.state = {
            loadingState: types_1.LoadingState.Loading,
            selectedTabIndex: 0,
        };
    }
    componentDidMount() {
        this.loadData();
    }
    render() {
        const { selectedTabIndex } = this.state;
        const tabs = this.tabs;
        return (react_1.default.createElement(eui_1.EuiTabbedContent, { tabs: tabs, onTabClick: this.onTabClick, selectedTab: tabs[selectedTabIndex] }));
    }
    get tabs() {
        const { intl } = this.props;
        const { loadingState, checkupData } = this.state;
        const commonProps = {
            loadingState,
            refreshCheckupData: this.loadData,
            setSelectedTabIndex: this.setSelectedTabIndex,
            // Remove this in last minor of the current major (eg. 6.7)
            alertBanner: react_1.default.createElement(latest_minor_banner_1.LatestMinorBanner, null),
        };
        return [
            {
                id: 'overview',
                name: intl.formatMessage({
                    id: 'xpack.upgradeAssistant.overviewTab.overviewTabTitle',
                    defaultMessage: 'Overview',
                }),
                content: react_1.default.createElement(overview_1.OverviewTab, Object.assign({ checkupData: checkupData }, commonProps)),
            },
            {
                id: 'cluster',
                name: intl.formatMessage({
                    id: 'xpack.upgradeAssistant.checkupTab.clusterTabLabel',
                    defaultMessage: 'Cluster',
                }),
                content: (react_1.default.createElement(checkup_1.CheckupTab, Object.assign({ key: "cluster", deprecations: checkupData ? checkupData.cluster : undefined, checkupLabel: intl.formatMessage({
                        id: 'xpack.upgradeAssistant.tabs.checkupTab.clusterLabel',
                        defaultMessage: 'cluster',
                    }) }, commonProps))),
            },
            {
                id: 'indices',
                name: intl.formatMessage({
                    id: 'xpack.upgradeAssistant.checkupTab.indicesTabLabel',
                    defaultMessage: 'Indices',
                }),
                content: (react_1.default.createElement(checkup_1.CheckupTab, Object.assign({ key: "indices", deprecations: checkupData ? checkupData.indices : undefined, checkupLabel: intl.formatMessage({
                        id: 'xpack.upgradeAssistant.checkupTab.indexLabel',
                        defaultMessage: 'index',
                    }), showBackupWarning: true }, commonProps))),
            },
        ];
    }
}
exports.UpgradeAssistantTabsUI = UpgradeAssistantTabsUI;
exports.UpgradeAssistantTabs = react_2.injectI18n(UpgradeAssistantTabsUI);
