// x-pack/plugins/beats_management/common/domain_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FilebeatModuleName;
(function (FilebeatModuleName) {
    FilebeatModuleName["system"] = "system";
    FilebeatModuleName["apache2"] = "apache2";
    FilebeatModuleName["nginx"] = "nginx";
    FilebeatModuleName["mongodb"] = "mongodb";
    FilebeatModuleName["elasticsearch"] = "elasticsearch";
})(FilebeatModuleName = exports.FilebeatModuleName || (exports.FilebeatModuleName = {}));
var MetricbeatModuleName;
(function (MetricbeatModuleName) {
    MetricbeatModuleName["system"] = "system";
    MetricbeatModuleName["apache2"] = "apache2";
    MetricbeatModuleName["nginx"] = "nginx";
    MetricbeatModuleName["mongodb"] = "mongodb";
    MetricbeatModuleName["elasticsearch"] = "elasticsearch";
})(MetricbeatModuleName = exports.MetricbeatModuleName || (exports.MetricbeatModuleName = {}));
var OutputType;
(function (OutputType) {
    OutputType["elasticsearch"] = "elasticsearch";
    OutputType["logstash"] = "logstash";
    OutputType["kafka"] = "kafka";
    OutputType["redis"] = "redis";
})(OutputType = exports.OutputType || (exports.OutputType = {}));
// TODO convery above enum to an array dynamicly here
exports.OutputTypesArray = ['elasticsearch', 'logstash', 'kafka', 'redis'];
