// x-pack/plugins/infra/server/lib/adapters/metrics/models/container/container_diskio_ops.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const adapter_types_1 = require("../../adapter_types");
exports.containerDiskIOOps = (timeField, indexPattern, interval) => ({
    id: 'containerDiskIOOps',
    requires: ['docker.disk'],
    index_pattern: indexPattern,
    interval,
    time_field: timeField,
    type: 'timeseries',
    series: [
        {
            id: 'read',
            split_mode: 'everything',
            metrics: [
                {
                    field: 'docker.diskio.read.ops',
                    id: 'avg-diskio-ops',
                    type: adapter_types_1.InfraMetricModelMetricType.avg,
                },
            ],
        },
        {
            id: 'write',
            split_mode: 'everything',
            metrics: [
                {
                    field: 'docker.diskio.write.ops',
                    id: 'avg-diskio-ops',
                    type: adapter_types_1.InfraMetricModelMetricType.avg,
                },
            ],
        },
    ],
});
