// x-pack/plugins/canvas/canvas_plugin/functions/common/dropdownControl.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dropdownControl = undefined;

var _lodash = require('lodash');

var dropdownControl = exports.dropdownControl = function dropdownControl() {
  return {
    name: 'dropdownControl',
    aliases: [],
    type: 'render',
    context: {
      types: ['datatable']
    },
    help: 'Configure a drop down filter control element',
    args: {
      filterColumn: {
        type: ['string'],
        help: 'The column or field to attach the filter to'
      },
      valueColumn: {
        type: ['string'],
        help: 'The datatable column from which to extract the unique values for the drop down'
      }
    },
    fn: function fn(context, _ref) {
      var valueColumn = _ref.valueColumn,
          filterColumn = _ref.filterColumn;

      var choices = [];
      if (context.rows[0][valueColumn]) {
        choices = (0, _lodash.uniq)(context.rows.map(function (row) {
          return row[valueColumn];
        })).sort();
      }

      var column = filterColumn || valueColumn;

      return {
        type: 'render',
        as: 'dropdown_filter',
        value: {
          column: column,
          choices: choices
        }
      };
    }
  };
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
