// x-pack/plugins/canvas/canvas_plugin/functions/common/formatnumber.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatnumber = undefined;

var _numeral = require('@elastic/numeral');

var _numeral2 = _interopRequireDefault(_numeral);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var formatnumber = exports.formatnumber = function formatnumber() {
  return {
    name: 'formatnumber',
    type: 'string',
    help: 'Turn a number into a string using a NumberJS format',
    context: {
      types: ['number']
    },
    args: {
      format: {
        aliases: ['_'],
        types: ['string'],
        help: 'NumeralJS format string http://numeraljs.com/#format'
      }
    },
    fn: function fn(context, args) {
      if (!args.format) {
        return String(context);
      }
      return (0, _numeral2.default)(context).format(args.format);
    }
  };
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
