// x-pack/plugins/canvas/canvas_plugin/functions/common/formatdate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatdate = undefined;

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var formatdate = exports.formatdate = function formatdate() {
  return {
    name: 'formatdate',
    type: 'string',
    help: 'Output a ms since epoch number as a formatted string',
    context: {
      types: ['number']
    },
    args: {
      format: {
        aliases: ['_'],
        types: ['string'],
        help: 'MomentJS Format with which to bucket (See https://momentjs.com/docs/#/displaying/)'
      }
    },
    fn: function fn(context, args) {
      if (!args.format) {
        return _moment2.default.utc(new Date(context)).toISOString();
      }
      return _moment2.default.utc(new Date(context)).format(args.format);
    }
  };
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
