// x-pack/plugins/canvas/canvas_plugin/functions/common/gte.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

var gte = exports.gte = function gte() {
  return {
    name: 'gte',
    type: 'boolean',
    help: 'Return if the context is greater than or equal to the argument',
    args: {
      value: {
        aliases: ['_'],
        types: ['boolean', 'number', 'string', 'null'],
        required: true,
        help: 'The value to compare the context to'
      }
    },
    fn: function fn(context, args) {
      if ((typeof context === 'undefined' ? 'undefined' : _typeof(context)) !== _typeof(args.value)) {
        return false;
      }
      return context >= args.value;
    }
  };
};
