// x-pack/plugins/canvas/canvas_plugin/functions/common/palette.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.palette = undefined;

var _palettes = require('../../../common/lib/palettes');

var palette = exports.palette = function palette() {
  return {
    name: 'palette',
    aliases: [],
    type: 'palette',
    help: 'Create a color palette',
    context: {
      types: ['null']
    },
    args: {
      color: {
        aliases: ['_'],
        multi: true,
        types: ['string'],
        help: 'Palette colors, rgba, hex, or HTML color string. Pass this multiple times.'
      },
      gradient: {
        types: ['boolean'],
        default: false,
        help: 'Prefer to make a gradient where supported and useful?',
        options: [true, false]
      },
      reverse: {
        type: ['boolean'],
        default: false,
        help: 'Reverse the palette',
        options: [true, false]
      }
    },
    fn: function fn(context, args) {
      var colors = [].concat(args.color || _palettes.palettes.paul_tor_14.colors);
      return {
        type: 'palette',
        colors: args.reverse ? colors.reverse() : colors,
        gradient: args.gradient
      };
    }
  };
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
