// x-pack/plugins/canvas/canvas_plugin/functions/common/revealImage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.revealImage = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /*
                                                                                                                                                                                                                                                                   * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
                                                                                                                                                                                                                                                                   * or more contributor license agreements. Licensed under the Elastic License;
                                                                                                                                                                                                                                                                   * you may not use this file except in compliance with the Elastic License.
                                                                                                                                                                                                                                                                   */

var _resolve_dataurl = require('../../../common/lib/resolve_dataurl');

var _elastic_outline = require('../../lib/elastic_outline');

var revealImage = exports.revealImage = function revealImage() {
  return {
    name: 'revealImage',
    aliases: [],
    type: 'render',
    help: 'Configure a image reveal element',
    context: {
      types: ['number']
    },
    args: {
      image: {
        types: ['string', 'null'],
        help: 'The image to reveal',
        default: _elastic_outline.elasticOutline
      },
      emptyImage: {
        types: ['string', 'null'],
        help: 'An optional background image to reveal over',
        default: null
      },
      origin: {
        types: ['string'],
        help: 'Where to start from. Eg, top, left, bottom or right',
        default: 'bottom',
        options: ['top', 'left', 'bottom', 'right']
      }
    },
    fn: function fn(percent, args) {
      if (percent > 1 || percent < 0) {
        throw new Error('Invalid value: \'' + percent + '\'. Percentage must be between 0 and 1');
      }

      return {
        type: 'render',
        as: 'revealImage',
        value: _extends({
          percent: percent
        }, args, {
          image: (0, _resolve_dataurl.resolveWithMissingImage)(args.image, _elastic_outline.elasticOutline),
          emptyImage: (0, _resolve_dataurl.resolveWithMissingImage)(args.emptyImage)
        })
      };
    }
  };
};
