// x-pack/plugins/canvas/canvas_plugin/functions/common/rounddate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rounddate = undefined;

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var rounddate = exports.rounddate = function rounddate() {
  return {
    name: 'rounddate',
    type: 'number',
    help: 'Round ms since epoch using a moment formatting string. Returns ms since epoch',
    context: {
      types: ['number']
    },
    args: {
      format: {
        aliases: ['_'],
        types: ['string'],
        help: 'MomentJS Format with which to bucket (See https://momentjs.com/docs/#/displaying/). For example "YYYY-MM" would round to the month'
      }
    },
    fn: function fn(context, args) {
      if (!args.format) {
        return context;
      }
      return _moment2.default.utc(_moment2.default.utc(context).format(args.format), args.format).valueOf();
    }
  };
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
