// x-pack/plugins/beats_management/public/components/config_list.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
// @ts-ignore
const eui_1 = require("@elastic/eui");
const react_1 = require("@kbn/i18n/react");
const react_2 = tslib_1.__importDefault(require("react"));
const config_schemas_translations_map_1 = require("../config_schemas_translations_map");
const ConfigListUi = props => (react_2.default.createElement(eui_1.EuiBasicTable, { items: props.configs || [], columns: [
        {
            field: 'type',
            name: props.intl.formatMessage({
                id: 'xpack.beatsManagement.tagTable.typeColumnName',
                defaultMessage: 'Type',
            }),
            truncateText: false,
            render: (value, config) => {
                const type = config_schemas_translations_map_1.getSupportedConfig().find((sc) => sc.value === config.type);
                return (react_2.default.createElement(eui_1.EuiLink, { onClick: () => props.onConfigClick('edit', config) }, type ? type.text : config.type));
            },
        },
        {
            field: 'module',
            name: props.intl.formatMessage({
                id: 'xpack.beatsManagement.tagTable.moduleColumnName',
                defaultMessage: 'Module',
            }),
            truncateText: false,
            render: (value) => {
                return (value ||
                    props.intl.formatMessage({
                        id: 'xpack.beatsManagement.tagTable.moduleColumn.notAvailibaleLabel',
                        defaultMessage: 'N/A',
                    }));
            },
        },
        {
            field: 'description',
            name: props.intl.formatMessage({
                id: 'xpack.beatsManagement.tagTable.descriptionColumnName',
                defaultMessage: 'Description',
            }),
        },
        {
            name: props.intl.formatMessage({
                id: 'xpack.beatsManagement.tagTable.actionsColumnName',
                defaultMessage: 'Actions',
            }),
            actions: [
                {
                    name: props.intl.formatMessage({
                        id: 'xpack.beatsManagement.tagTable.actions.removeButtonAriaLabel',
                        defaultMessage: 'Remove',
                    }),
                    description: props.intl.formatMessage({
                        id: 'xpack.beatsManagement.tagTable.actions.removeTooltip',
                        defaultMessage: 'Remove this config from tag',
                    }),
                    type: 'icon',
                    icon: 'trash',
                    onClick: (item) => props.onConfigClick('delete', item),
                },
            ],
        },
    ] }));
exports.ConfigList = react_1.injectI18n(ConfigListUi);
