// x-pack/plugins/apm/public/components/app/ServiceOverview/ServiceList/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importDefault(require("react"));
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
const variables_1 = require("../../../../style/variables");
const formatters_1 = require("../../../../utils/formatters");
const url_1 = require("../../../../utils/url");
const ManagedTable_1 = require("../../../shared/ManagedTable");
function formatNumber(value) {
    if (value === 0) {
        return '0';
    }
    else if (value <= 0.1) {
        return '< 0.1';
    }
    else {
        return formatters_1.asDecimal(value);
    }
}
function formatString(value) {
    return value || 'N/A';
}
const AppLink = styled_components_1.default(url_1.RelativeLink) `
  font-size: ${variables_1.fontSizes.large};
  ${variables_1.truncate('100%')};
`;
exports.SERVICE_COLUMNS = [
    {
        field: 'serviceName',
        name: 'Name',
        width: '50%',
        sortable: true,
        render: (serviceName) => (react_1.default.createElement(eui_1.EuiToolTip, { content: formatString(serviceName), id: "service-name-tooltip" },
            react_1.default.createElement(AppLink, { path: `/${serviceName}/transactions` }, formatString(serviceName))))
    },
    {
        field: 'agentName',
        name: 'Agent',
        sortable: true,
        render: (agentName) => formatString(agentName)
    },
    {
        field: 'avgResponseTime',
        name: 'Avg. response time',
        sortable: true,
        dataType: 'number',
        render: (value) => formatters_1.asMillis(value)
    },
    {
        field: 'transactionsPerMinute',
        name: 'Trans. per minute',
        sortable: true,
        dataType: 'number',
        render: (value) => `${formatNumber(value)} tpm`
    },
    {
        field: 'errorsPerMinute',
        name: 'Errors per minute',
        sortable: true,
        dataType: 'number',
        render: (value) => `${formatNumber(value)} err.`
    }
];
function ServiceList({ items = [], noItemsMessage }) {
    return (react_1.default.createElement(ManagedTable_1.ManagedTable, { columns: exports.SERVICE_COLUMNS, items: items, noItemsMessage: noItemsMessage, initialSort: { field: 'serviceName', direction: 'asc' } }));
}
exports.ServiceList = ServiceList;
