// x-pack/plugins/apm/server/lib/metrics/get_memory_chart_data/fetcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("x-pack/plugins/apm/common/constants");
const metricsFetcher_1 = require("../metricsFetcher");
const percentSystemMemoryUsedScript = {
    script: {
        lang: 'expression',
        source: `1 - doc['${constants_1.METRIC_SYSTEM_FREE_MEMORY}'] / doc['${constants_1.METRIC_SYSTEM_TOTAL_MEMORY}']`
    }
};
async function fetch(args) {
    return metricsFetcher_1.fetchMetrics({
        ...args,
        bool: {
            must: [
                {
                    exists: {
                        field: constants_1.METRIC_SYSTEM_TOTAL_MEMORY
                    }
                },
                {
                    exists: {
                        field: constants_1.METRIC_SYSTEM_FREE_MEMORY
                    }
                }
            ]
        },
        timeseriesBucketAggregations: {
            averagePercentMemoryUsed: { avg: percentSystemMemoryUsedScript },
            maximumPercentMemoryUsed: { max: percentSystemMemoryUsedScript }
        },
        totalAggregations: {
            averagePercentMemoryUsed: { avg: percentSystemMemoryUsedScript },
            maximumPercentMemoryUsed: { max: percentSystemMemoryUsedScript }
        }
    });
}
exports.fetch = fetch;
