// infra/server/lib/domains/log_entries_domain/builtin_rules/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const filebeat_apache2_1 = require("./filebeat_apache2");
const filebeat_nginx_1 = require("./filebeat_nginx");
const filebeat_redis_1 = require("./filebeat_redis");
const filebeat_system_1 = require("./filebeat_system");
const generic_1 = require("./generic");
exports.builtinRules = [
    ...filebeat_apache2_1.filebeatApache2Rules,
    ...filebeat_nginx_1.filebeatNginxRules,
    ...filebeat_redis_1.filebeatRedisRules,
    ...filebeat_system_1.filebeatSystemRules,
    ...generic_1.genericRules,
    {
        when: {
            exists: ['source'],
        },
        format: [
            {
                constant: 'failed to format message from ',
            },
            {
                field: 'source',
            },
        ],
    },
    {
        when: {
            exists: [],
        },
        format: [
            {
                constant: 'failed to find message',
            },
        ],
    },
];
