// x-pack/plugins/infra/server/lib/adapters/sources/configuration_sources_adapter.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class InfraConfigurationSourcesAdapter {
    constructor(configuration) {
        this.configuration = configuration;
    }
    async getAll() {
        const sourceConfigurations = (await this.configuration.get()).sources || {
            default: DEFAULT_SOURCE,
        };
        const sourceConfigurationsWithDefault = {
            ...sourceConfigurations,
            default: {
                ...DEFAULT_SOURCE,
                ...(sourceConfigurations.default || {}),
            },
        };
        return Object.entries(sourceConfigurationsWithDefault).reduce((result, [sourceId, sourceConfiguration]) => ({
            ...result,
            [sourceId]: {
                ...sourceConfiguration,
                fields: {
                    ...DEFAULT_FIELDS,
                    ...(sourceConfiguration.fields || {}),
                },
            },
        }), {});
    }
}
exports.InfraConfigurationSourcesAdapter = InfraConfigurationSourcesAdapter;
const DEFAULT_FIELDS = {
    container: 'docker.container.id',
    host: 'host.name',
    message: ['message', '@message'],
    pod: 'kubernetes.pod.uid',
    tiebreaker: '_doc',
    timestamp: '@timestamp',
};
const DEFAULT_SOURCE = {
    metricAlias: 'metricbeat-*',
    logAlias: 'filebeat-*',
    fields: DEFAULT_FIELDS,
};
