// x-pack/plugins/canvas/canvas_plugin/functions/common/plot/series_style_to_flot.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.seriesStyleToFlot = undefined;

var _lodash = require('lodash');

var seriesStyleToFlot = exports.seriesStyleToFlot = function seriesStyleToFlot(seriesStyle) {
  if (!seriesStyle) {
    return {};
  }

  var lines = (0, _lodash.get)(seriesStyle, 'lines');
  var bars = (0, _lodash.get)(seriesStyle, 'bars');
  var fill = (0, _lodash.get)(seriesStyle, 'fill');
  var color = (0, _lodash.get)(seriesStyle, 'color');
  var stack = (0, _lodash.get)(seriesStyle, 'stack');
  var horizontal = (0, _lodash.get)(seriesStyle, 'horizontalBars', false);

  var flotStyle = {
    numbers: {
      show: true
    },
    lines: {
      show: lines > 0,
      lineWidth: lines,
      fillColor: color,
      fill: fill / 10
    },
    bars: {
      show: bars > 0,
      barWidth: bars,
      fill: 1,
      align: 'center',
      horizontal: horizontal
    },
    // This is here intentionally even though it is the default.
    // We use the `size` plugins for this and if the user says they want points
    // we just set the size to be static.
    points: { show: false },
    bubbles: {
      fill: fill
    }
  };

  if (stack) {
    flotStyle.stack = stack;
  }
  if (color) {
    flotStyle.color = color;
  }

  return flotStyle;
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
