// x-pack/plugins/canvas/canvas_plugin/functions/server/src/demodata/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.demodata = undefined;

var _lodash = require('lodash');

var _query = require('../../../../../common/lib/datatable/query');

var _get_demo_rows = require('./get_demo_rows');

var demodata = exports.demodata = function demodata() {
  return {
    name: 'demodata',
    aliases: [],
    type: 'datatable',
    help: 'A mock data set that includes project CI times with usernames, countries and run phases',
    context: {
      types: ['filter']
    },
    args: {
      type: {
        types: ['string', 'null'],
        aliases: ['_'],
        help: 'The name of the demo data set to use',
        default: 'ci',
        options: ['ci', 'shirts']
      }
    },
    fn: function fn(context, args) {
      var demoRows = (0, _get_demo_rows.getDemoRows)(args.type);

      var set = {};
      if (args.type === 'ci') {
        set = {
          columns: [{ name: 'time', type: 'date' }, { name: 'cost', type: 'number' }, { name: 'username', type: 'string' }, { name: 'price', type: 'number' }, { name: 'age', type: 'number' }, { name: 'country', type: 'string' }, { name: 'state', type: 'string' }, { name: 'project', type: 'string' }, { name: 'percent_uptime', type: 'number' }],
          rows: (0, _lodash.sortBy)(demoRows, 'time')
        };
      } else if (args.type === 'shirts') {
        set = {
          columns: [{ name: 'size', type: 'string' }, { name: 'color', type: 'string' }, { name: 'price', type: 'number' }, { name: 'cut', type: 'string' }],
          rows: demoRows
        };
      }

      var _set = set,
          columns = _set.columns,
          rows = _set.rows;

      return (0, _query.queryDatatable)({
        type: 'datatable',
        columns: columns,
        rows: rows
      }, context);
    }
  };
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
