// x-pack/plugins/beats_management/public/pages/overview/enrolled_beats.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const i18n_1 = require("@kbn/i18n");
const react_1 = require("@kbn/i18n/react");
const lodash_1 = require("lodash");
const moment_1 = tslib_1.__importDefault(require("moment"));
const react_2 = tslib_1.__importDefault(require("react"));
const constants_1 = require("x-pack/plugins/beats_management/common/constants");
const enroll_beats_1 = require("../../components/enroll_beats");
const breadcrumb_1 = require("../../components/navigation/breadcrumb");
const table_1 = require("../../components/table");
const action_schema_1 = require("../../components/table/action_schema");
const table_2 = require("../../components/table/table");
const with_kuery_autocompletion_1 = require("../../containers/with_kuery_autocompletion");
class BeatsPageComponent extends react_2.default.PureComponent {
    constructor(props) {
        super(props);
        this.tableRef = react_2.default.createRef();
        this.renderActionArea = () => (react_2.default.createElement(react_2.default.Fragment, null,
            react_2.default.createElement(eui_1.EuiButtonEmpty, { onClick: () => {
                    // random, but specific number ensures new tab does not overwrite another _newtab in chrome
                    // and at the same time not truly random so that many clicks of the link open many tabs at this same URL
                    window.open('https://www.elastic.co/guide/en/beats/libbeat/current/getting-started.html', '_newtab35628937456');
                } },
                react_2.default.createElement(react_1.FormattedMessage, { id: "xpack.beatsManagement.beats.installBeatsLearningButtonLabel", defaultMessage: "Learn how to install beats" })),
            react_2.default.createElement(eui_1.EuiButton, { size: "s", color: "primary", onClick: async () => {
                    this.props.goTo(`/overview/enrolled_beats/enroll`);
                } },
                react_2.default.createElement(react_1.FormattedMessage, { id: "xpack.beatsManagement.beats.enrollBeatsButtonLabel", defaultMessage: "Enroll Beats" })),
            this.props.location.pathname === '/overview/enrolled_beats/enroll' && (react_2.default.createElement(eui_1.EuiOverlayMask, null,
                react_2.default.createElement(eui_1.EuiModal, { onClose: () => {
                        this.props.setUrlState({
                            enrollmentToken: '',
                        });
                        this.props.goTo(`/overview/enrolled_beats`);
                    }, style: { width: '640px' } },
                    react_2.default.createElement(eui_1.EuiModalHeader, null,
                        react_2.default.createElement(eui_1.EuiModalHeaderTitle, null,
                            react_2.default.createElement(react_1.FormattedMessage, { id: "xpack.beatsManagement.beats.enrollNewBeatsTitle", defaultMessage: "Enroll a new Beat" }))),
                    react_2.default.createElement(eui_1.EuiModalBody, null,
                        react_2.default.createElement(enroll_beats_1.EnrollBeat, { frameworkBasePath: this.props.libs.framework.info.basePath, enrollmentToken: this.props.urlState.enrollmentToken, getBeatWithToken: this.props.containers.beats.getBeatWithToken, createEnrollmentToken: async () => {
                                const enrollmentToken = await this.props.libs.tokens.createEnrollmentToken();
                                this.props.setUrlState({
                                    enrollmentToken,
                                });
                            }, onBeatEnrolled: () => {
                                this.props.setUrlState({
                                    enrollmentToken: '',
                                });
                            } }),
                        !this.props.urlState.enrollmentToken && (react_2.default.createElement(react_2.default.Fragment, null,
                            react_2.default.createElement(eui_1.EuiButton, { size: "s", color: "primary", style: { marginLeft: 10 }, onClick: async () => {
                                    this.props.goTo('/overview/enrolled_beats');
                                } }, "Done")))))))));
        this.notifyBeatDisenrolled = async (beats) => {
            const { intl } = this.props;
            let title;
            let text;
            if (beats.length === 1) {
                title = intl.formatMessage({
                    id: 'xpack.beatsManagement.beats.beatDisenrolledNotificationTitle',
                    defaultMessage: '{firstBeatNameOrId} disenrolled',
                }, {
                    firstBeatNameOrId: `"${beats[0].name || beats[0].id}"`,
                });
                text = intl.formatMessage({
                    id: 'xpack.beatsManagement.beats.beatDisenrolledNotificationDescription',
                    defaultMessage: 'Beat with ID {firstBeatId} was disenrolled.',
                }, {
                    firstBeatId: `"${beats[0].id}"`,
                });
            }
            else {
                title = intl.formatMessage({
                    id: 'xpack.beatsManagement.beats.disenrolledBeatsNotificationTitle',
                    defaultMessage: '{beatsLength} beats disenrolled',
                }, {
                    beatsLength: beats.length,
                });
            }
            this.setState({
                notifications: this.state.notifications.concat({
                    color: 'warning',
                    id: `disenroll_${new Date()}`,
                    title,
                    text,
                }),
            });
        };
        this.notifyUpdatedTagAssociation = (action, beats, tag) => {
            const { intl } = this.props;
            const notificationMessage = action === 'removed'
                ? intl.formatMessage({
                    id: 'xpack.beatsManagement.beats.removedNotificationDescription',
                    defaultMessage: 'Removed tag {tag} from {assignmentsLength, plural, one {beat {beatName}} other {# beats}}.',
                }, {
                    tag: `"${tag}"`,
                    assignmentsLength: beats.length,
                    beatName: `"${beats[0].name || beats[0].id}"`,
                })
                : intl.formatMessage({
                    id: 'xpack.beatsManagement.beats.addedNotificationDescription',
                    defaultMessage: 'Added tag {tag} to {assignmentsLength, plural, one {beat {beatName}} other {# beats}}.',
                }, {
                    tag: `"${tag}"`,
                    assignmentsLength: beats.length,
                    beatName: `"${beats[0].name || beats[0].id}"`,
                });
            const notificationTitle = action === 'removed'
                ? intl.formatMessage({
                    id: 'xpack.beatsManagement.beats.removedNotificationTitle',
                    defaultMessage: '{assignmentsLength, plural, one {Tag} other {Tags}} removed',
                }, {
                    assignmentsLength: beats.length,
                })
                : intl.formatMessage({
                    id: 'xpack.beatsManagement.beats.addedNotificationTitle',
                    defaultMessage: '{assignmentsLength, plural, one {Tag} other {Tags}} added',
                }, {
                    assignmentsLength: beats.length,
                });
            this.setState({
                notifications: this.state.notifications.concat({
                    color: 'success',
                    id: `tag-${moment_1.default.now()}`,
                    text: react_2.default.createElement("p", null, notificationMessage),
                    title: notificationTitle,
                }),
            });
        };
        this.getSelectedBeats = () => {
            if (!this.tableRef.current) {
                return [];
            }
            const selectedIds = this.tableRef.current.state.selection.map((beat) => beat.id);
            const beats = [];
            selectedIds.forEach((id) => {
                const beat = this.props.containers.beats.state.list.find(b => b.id === id);
                if (beat) {
                    beats.push(beat);
                }
            });
            return beats;
        };
        this.filterTags = (tags) => {
            return this.selectedBeatConfigsRequireUniqueness()
                ? tags.map(this.disableTagForUniquenessEnforcement)
                : tags;
        };
        this.configBlocksRequireUniqueness = (configurationBlocks) => lodash_1.intersection(constants_1.UNIQUENESS_ENFORCING_TYPES, configurationBlocks.map(block => block.type))
            .length !== 0;
        this.disableTagForUniquenessEnforcement = (tag) => this.configBlocksRequireUniqueness(tag.configuration_blocks) &&
            // if > 0 beats are associated with the tag, it will result in disassociation, so do not disable it
            !this.getSelectedBeats().some(beat => beat.full_tags.some(({ id }) => id === tag.id))
            ? { ...tag, disabled: true }
            : tag;
        this.selectedBeatConfigsRequireUniqueness = () => 
        // union beat tags
        lodash_1.flatten(this.getSelectedBeats().map(({ full_tags }) => full_tags))
            // map tag list to bool
            .map(tag => {
            return this.configBlocksRequireUniqueness(tag ? tag.configuration_blocks : []);
        })
            // reduce to result
            .reduce((acc, cur) => acc || cur, false);
        this.state = {
            notifications: [],
            tags: null,
        };
        if (props.urlState.beatsKBar) {
            props.containers.beats.reload(props.urlState.beatsKBar);
        }
        props.renderAction(this.renderActionArea);
    }
    render() {
        return (react_2.default.createElement(react_2.default.Fragment, null,
            react_2.default.createElement(breadcrumb_1.Breadcrumb, { title: i18n_1.i18n.translate('xpack.beatsManagement.breadcrumb.enrolledBeats', {
                    defaultMessage: 'Enrolled Beats',
                }), path: `/overview/enrolled_beats` }),
            react_2.default.createElement(with_kuery_autocompletion_1.WithKueryAutocompletion, { libs: this.props.libs, fieldPrefix: "beat" }, autocompleteProps => (react_2.default.createElement(table_1.Table, { kueryBarProps: {
                    ...autocompleteProps,
                    filterQueryDraft: 'false',
                    isValid: this.props.libs.elasticsearch.isKueryValid(this.props.urlState.beatsKBar || ''),
                    onChange: (value) => {
                        this.props.setUrlState({ beatsKBar: value });
                        this.props.containers.beats.reload(this.props.urlState.beatsKBar);
                    },
                    onSubmit: () => null,
                    value: this.props.urlState.beatsKBar || '',
                }, actions: action_schema_1.beatsListActions, actionData: {
                    tags: this.filterTags(this.props.containers.tags.state.list),
                }, actionHandler: async (action, payload) => {
                    switch (action) {
                        case table_2.AssignmentActionType.Assign:
                            const status = await this.props.containers.beats.toggleTagAssignment(payload, this.getSelectedBeats());
                            this.notifyUpdatedTagAssociation(status, this.getSelectedBeats(), payload);
                            break;
                        case table_2.AssignmentActionType.Delete:
                            this.props.containers.beats.deactivate(this.getSelectedBeats());
                            this.notifyBeatDisenrolled(this.getSelectedBeats());
                            break;
                        case table_2.AssignmentActionType.Reload:
                            this.props.containers.tags.reload();
                            break;
                    }
                }, items: lodash_1.sortBy(this.props.containers.beats.state.list, 'id') || [], ref: this.tableRef, type: table_1.BeatsTableType }))),
            react_2.default.createElement(eui_1.EuiGlobalToastList, { toasts: this.state.notifications, dismissToast: () => this.setState({ notifications: [] }), toastLifeTimeMs: 5000 })));
    }
}
exports.BeatsPage = react_1.injectI18n(BeatsPageComponent);
