// x-pack/plugins/canvas/canvas_plugin/functions/server/src/escount.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.escount = undefined;

var _build_es_request = require('../../../../server/lib/build_es_request');

var escount = exports.escount = function escount() {
  return {
    name: 'escount',
    type: 'number',
    help: 'Query elasticsearch for a count of the number of hits matching a query',
    context: {
      types: ['filter']
    },
    args: {
      index: {
        types: ['string', 'null'],
        default: '_all',
        help: 'Specify an index pattern. Eg "logstash-*"'
      },
      query: {
        types: ['string'],
        aliases: ['_', 'q'],
        help: 'A Lucene query string',
        default: '"-_index:.kibana"'
      }
    },
    fn: function fn(context, args, handlers) {
      context.and = context.and.concat([{
        type: 'luceneQueryString',
        query: args.query
      }]);

      var esRequest = (0, _build_es_request.buildESRequest)({
        index: args.index,
        body: {
          query: {
            bool: {
              must: [{ match_all: {} }]
            }
          }
        }
      }, context);

      return handlers.elasticsearchClient('count', esRequest).then(function (resp) {
        return resp.count;
      });
    }
  };
}; /*
    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
    * or more contributor license agreements. Licensed under the Elastic License;
    * you may not use this file except in compliance with the Elastic License.
    */
