// infra/server/lib/sources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
const i18n_1 = require("@kbn/i18n");
class InfraSources {
    constructor(adapter) {
        this.adapter = adapter;
    }
    async getConfiguration(sourceId) {
        const sourceConfigurations = await this.getAllConfigurations();
        const requestedSourceConfiguration = sourceConfigurations[sourceId];
        if (!requestedSourceConfiguration) {
            throw new Error(i18n_1.i18n.translate('xpack.infra.infraSources.failedToFindSourceErrorMessage', {
                defaultMessage: 'Failed to find source {sourceId}',
                values: {
                    sourceId: `'${sourceId}'`,
                },
            }));
        }
        return requestedSourceConfiguration;
    }
    getAllConfigurations() {
        return this.adapter.getAll();
    }
}
exports.InfraSources = InfraSources;
