// x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importStar(require("react"));
const notify_1 = require("ui/notify");
const ml_1 = require("x-pack/plugins/apm/public/services/rest/ml");
const savedObjects_1 = require("x-pack/plugins/apm/public/services/rest/savedObjects");
const machineLearningJobs_1 = require("x-pack/plugins/apm/public/store/reactReduxRequest/machineLearningJobs");
const url_1 = require("x-pack/plugins/apm/public/utils/url");
const TransactionSelect_1 = require("./TransactionSelect");
class MachineLearningFlyout extends react_1.Component {
    constructor() {
        super(...arguments);
        this.state = {
            isLoading: false,
            hasIndexPattern: false,
            hasMLJob: false,
            selectedTransactionType: this.props.urlParams.transactionType
        };
        this.createJob = async () => {
            this.setState({ isLoading: true });
            try {
                const { serviceName, transactionType } = this.props.urlParams;
                if (serviceName) {
                    const res = await ml_1.startMLJob({ serviceName, transactionType });
                    const didSucceed = res.datafeeds[0].success && res.jobs[0].success;
                    if (!didSucceed) {
                        throw new Error('Creating ML job failed');
                    }
                    this.addSuccessToast();
                }
            }
            catch (e) {
                this.addErrorToast();
            }
            this.setState({ isLoading: false });
            this.props.onClose();
        };
        this.addErrorToast = () => {
            const { urlParams } = this.props;
            const { serviceName = 'unknown' } = urlParams;
            if (!serviceName) {
                return;
            }
            notify_1.toastNotifications.addWarning({
                title: 'Job creation failed',
                text: (react_1.default.createElement("p", null, "Your current license may not allow for creating machine learning jobs, or this job may already exist."))
            });
        };
        this.addSuccessToast = () => {
            const { location, urlParams } = this.props;
            const { serviceName = 'unknown', transactionType } = urlParams;
            notify_1.toastNotifications.addSuccess({
                title: 'Job successfully created',
                text: (react_1.default.createElement("p", null,
                    "The analysis is now running for ",
                    serviceName,
                    " (",
                    transactionType,
                    "). It might take a while before results are added to the response times graph.",
                    ' ',
                    react_1.default.createElement(url_1.ViewMLJob, { serviceName: serviceName, transactionType: transactionType, location: location }, "View job")))
            });
        };
    }
    async componentDidMount() {
        const indexPattern = await savedObjects_1.getAPMIndexPattern();
        this.setState({ hasIndexPattern: !!indexPattern });
    }
    componentDidUpdate(prevProps) {
        if (prevProps.urlParams.transactionType !==
            this.props.urlParams.transactionType) {
            this.setState({
                selectedTransactionType: this.props.urlParams.transactionType
            });
        }
    }
    render() {
        const { isOpen, onClose, urlParams } = this.props;
        const { serviceName, transactionType } = urlParams;
        const { isLoading, hasIndexPattern, selectedTransactionType } = this.state;
        if (!isOpen || !serviceName) {
            return null;
        }
        return (react_1.default.createElement(machineLearningJobs_1.MLJobsRequest, { serviceName: serviceName, render: ({ data, status }) => {
                if (status === 'LOADING') {
                    return null;
                }
                const hasMLJob = data.jobs.some(job => job.jobId &&
                    job.jobId.startsWith(ml_1.getMlPrefix(serviceName, selectedTransactionType)));
                return (react_1.default.createElement(eui_1.EuiFlyout, { onClose: onClose, size: "s" },
                    react_1.default.createElement(eui_1.EuiFlyoutHeader, null,
                        react_1.default.createElement(eui_1.EuiTitle, null,
                            react_1.default.createElement("h2", null, "Enable anomaly detection")),
                        react_1.default.createElement(eui_1.EuiSpacer, { size: "s" })),
                    react_1.default.createElement(eui_1.EuiFlyoutBody, null,
                        hasMLJob && (react_1.default.createElement("div", null,
                            react_1.default.createElement(eui_1.EuiCallOut, { title: "Job already exists", color: "success", iconType: "check" },
                                react_1.default.createElement("p", null,
                                    "There is currently a job running for ",
                                    serviceName,
                                    " (",
                                    transactionType,
                                    ").",
                                    ' ',
                                    react_1.default.createElement(url_1.ViewMLJob, { serviceName: serviceName, transactionType: transactionType, location: location }, "View existing job"))),
                            react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }))),
                        !hasIndexPattern && (react_1.default.createElement("div", null,
                            react_1.default.createElement(eui_1.EuiCallOut, { title: react_1.default.createElement("span", null,
                                    "No APM index pattern available. To create a job, please import the APM index pattern via the",
                                    ' ',
                                    react_1.default.createElement(url_1.KibanaLink, { pathname: '/app/kibana', hash: `/home/tutorial/apm` }, "Setup Instructions")), color: "warning", iconType: "alert" }),
                            react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }))),
                        react_1.default.createElement(eui_1.EuiText, null,
                            react_1.default.createElement("p", null,
                                "Here you can create a machine learning job to calculate anomaly scores on durations for APM transactions within the",
                                ' ',
                                serviceName,
                                " service. Once enabled,",
                                ' ',
                                react_1.default.createElement("b", null, "the transaction duration graph"),
                                " will show the expected bounds and annotate the graph once the anomaly score is >=75."),
                            react_1.default.createElement("p", null,
                                "Jobs can be created for each service + transaction type combination. Once a job is created, you can manage it and see more details in the",
                                ' ',
                                react_1.default.createElement(url_1.KibanaLink, { pathname: '/app/ml' }, "Machine Learning jobs management page"),
                                ".",
                                ' ',
                                react_1.default.createElement("em", null, "Note: It might take a few minutes for the job to begin calculating results."))),
                        react_1.default.createElement(eui_1.EuiSpacer, null)),
                    react_1.default.createElement(eui_1.EuiFlyoutFooter, null,
                        react_1.default.createElement(eui_1.EuiFlexGroup, { justifyContent: "spaceBetween", alignItems: "flexEnd" },
                            react_1.default.createElement(eui_1.EuiFlexItem, null, this.props.serviceTransactionTypes.length > 1 ? (react_1.default.createElement(TransactionSelect_1.TransactionSelect, { types: this.props.serviceTransactionTypes, selected: this.state.selectedTransactionType, existingJobs: data.jobs, onChange: value => this.setState({
                                    selectedTransactionType: value
                                }) })) : null),
                            react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                                react_1.default.createElement(eui_1.EuiFormRow, null,
                                    react_1.default.createElement(eui_1.EuiButton, { onClick: this.createJob, fill: true, disabled: isLoading || hasMLJob || !hasIndexPattern }, "Create new job")))))));
            } }));
    }
}
exports.MachineLearningFlyout = MachineLearningFlyout;
