// x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/WatcherFlyout.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const moment_timezone_1 = tslib_1.__importDefault(require("moment-timezone"));
const react_1 = tslib_1.__importStar(require("react"));
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
const chrome_1 = tslib_1.__importDefault(require("ui/chrome"));
const notify_1 = require("ui/notify");
const xpack_1 = require("../../../../utils/documentation/xpack");
const url_1 = require("../../../../utils/url");
const createErrorGroupWatch_1 = require("./createErrorGroupWatch");
const getUserTimezone = lodash_1.default.memoize(() => {
    const uiSettings = chrome_1.default.getUiSettingsClient();
    return uiSettings.get('dateFormat:tz') === 'Browser'
        ? moment_timezone_1.default.tz.guess()
        : uiSettings.get('dateFormat:tz');
});
const SmallInput = styled_components_1.default.div `
  .euiFormRow {
    max-width: 85px;
  }
  .euiFormHelpText {
    width: 200px;
  }
`;
class WatcherFlyout extends react_1.Component {
    constructor() {
        super(...arguments);
        this.state = {
            schedule: 'daily',
            threshold: 10,
            actions: {
                slack: false,
                email: false
            },
            interval: {
                value: 10,
                unit: 'm'
            },
            daily: '08:00',
            emails: '',
            slackUrl: ''
        };
        this.onChangeSchedule = (schedule) => {
            this.setState({ schedule });
        };
        this.onChangeThreshold = (event) => {
            this.setState({
                threshold: parseInt(event.target.value, 10)
            });
        };
        this.onChangeDailyUnit = (event) => {
            this.setState({
                daily: event.target.value
            });
        };
        this.onChangeIntervalValue = (event) => {
            this.setState({
                interval: {
                    value: parseInt(event.target.value, 10),
                    unit: this.state.interval.unit
                }
            });
        };
        this.onChangeIntervalUnit = (event) => {
            this.setState({
                interval: {
                    value: this.state.interval.value,
                    unit: event.target.value
                }
            });
        };
        this.onChangeAction = (actionName) => {
            this.setState({
                actions: {
                    ...this.state.actions,
                    [actionName]: !this.state.actions[actionName]
                }
            });
        };
        this.onChangeEmails = (event) => {
            this.setState({ emails: event.target.value });
        };
        this.onChangeSlackUrl = (event) => {
            this.setState({ slackUrl: event.target.value });
        };
        this.createWatch = () => {
            const { serviceName } = this.props.urlParams;
            if (!serviceName) {
                return;
            }
            const emails = this.state.actions.email
                ? this.state.emails
                    .split(',')
                    .map(email => email.trim())
                    .filter(email => !!email)
                : [];
            const slackUrl = this.state.actions.slack ? this.state.slackUrl : '';
            const schedule = this.state.schedule === 'interval'
                ? {
                    interval: `${this.state.interval.value}${this.state.interval.unit}`
                }
                : {
                    daily: { at: `${this.state.daily}` }
                };
            const timeRange = this.state.schedule === 'interval'
                ? {
                    value: this.state.interval.value,
                    unit: this.state.interval.unit
                }
                : {
                    value: 24,
                    unit: 'h'
                };
            return createErrorGroupWatch_1.createErrorGroupWatch({
                emails,
                schedule,
                serviceName,
                slackUrl,
                threshold: this.state.threshold,
                timeRange
            })
                .then((id) => {
                this.props.onClose();
                this.addSuccessToast(id);
            })
                .catch(e => {
                // tslint:disable-next-line
                console.error(e);
                this.addErrorToast();
            });
        };
        this.addErrorToast = () => {
            notify_1.toastNotifications.addWarning({
                title: 'Watch creation failed',
                text: react_1.default.createElement("p", null, "Make sure your user has permission to create watches.")
            });
        };
        this.addSuccessToast = (id) => {
            notify_1.toastNotifications.addSuccess({
                title: 'New watch created!',
                text: (react_1.default.createElement("p", null,
                    "The watch is now ready and will send error reports for",
                    ' ',
                    this.props.urlParams.serviceName,
                    ".",
                    ' ',
                    react_1.default.createElement(url_1.UnconnectedKibanaLink, { location: this.props.location, pathname: '/app/kibana', hash: `/management/elasticsearch/watcher/watches/watch/${id}` }, "View watch.")))
            });
        };
    }
    render() {
        if (!this.props.isOpen) {
            return null;
        }
        const userTimezoneSetting = getUserTimezone();
        const dailyTime = this.state.daily;
        const inputTime = `${dailyTime}Z`; // Add tz to make into UTC
        const inputFormat = 'HH:mmZ'; // Parse as 24 hour w. tz
        const dailyTimeFormatted = moment_timezone_1.default(inputTime, inputFormat)
            .tz(userTimezoneSetting)
            .format('HH:mm'); // Format as 24h
        const dailyTime12HourFormatted = moment_timezone_1.default(inputTime, inputFormat)
            .tz(userTimezoneSetting)
            .format('hh:mm A (z)'); // Format as 12h w. tz
        // Generate UTC hours for Daily Report select field
        const intervalHours = lodash_1.default.range(24).map(i => {
            const hour = lodash_1.default.padLeft(i.toString(), 2, '0');
            return { value: `${hour}:00`, text: `${hour}:00 UTC` };
        });
        const flyoutBody = (react_1.default.createElement(eui_1.EuiText, null,
            react_1.default.createElement("p", null,
                "This form will assist in creating a Watch that can notify you of error occurrences from this service. To learn more about Watcher, please read our",
                ' ',
                react_1.default.createElement(eui_1.EuiLink, { target: "_blank", href: xpack_1.XPACK_DOCS.xpackWatcher }, "documentation"),
                "."),
            react_1.default.createElement(eui_1.EuiForm, null,
                react_1.default.createElement("h3", null, "Condition"),
                react_1.default.createElement(eui_1.EuiFormRow, { label: "Occurrences threshold per error group", helpText: "Threshold to be met for error group to be included in report.", compressed: true },
                    react_1.default.createElement(eui_1.EuiFieldNumber, { icon: "number", min: 1, value: this.state.threshold, onChange: this.onChangeThreshold })),
                react_1.default.createElement("h3", null, "Trigger schedule"),
                react_1.default.createElement("p", null, "Choose the time interval for the report, when the threshold is exceeded."),
                react_1.default.createElement(eui_1.EuiRadio, { id: "daily", label: "Daily report", onChange: () => this.onChangeSchedule('daily'), checked: this.state.schedule === 'daily' }),
                react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }),
                react_1.default.createElement(eui_1.EuiFormRow, { helpText: `The daily report will be sent at ${dailyTimeFormatted} / ${dailyTime12HourFormatted}.`, compressed: true },
                    react_1.default.createElement(eui_1.EuiSelect, { value: dailyTime, onChange: this.onChangeDailyUnit, options: intervalHours, disabled: this.state.schedule !== 'daily' })),
                react_1.default.createElement(eui_1.EuiRadio, { id: "interval", label: "Interval", onChange: () => this.onChangeSchedule('interval'), checked: this.state.schedule === 'interval' }),
                react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }),
                react_1.default.createElement(eui_1.EuiFlexGroup, null,
                    react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                        react_1.default.createElement(SmallInput, null,
                            react_1.default.createElement(eui_1.EuiFormRow, { helpText: "Time interval between reports.", compressed: true },
                                react_1.default.createElement(eui_1.EuiFieldNumber, { icon: "clock", min: 1, value: this.state.interval.value, onChange: this.onChangeIntervalValue, disabled: this.state.schedule !== 'interval' })))),
                    react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                        react_1.default.createElement(eui_1.EuiFormRow, { compressed: true },
                            react_1.default.createElement(eui_1.EuiSelect, { value: this.state.interval.unit, onChange: this.onChangeIntervalUnit, options: [
                                    {
                                        value: 'm',
                                        text: 'mins'
                                    },
                                    {
                                        value: 'h',
                                        text: 'hrs'
                                    }
                                ], disabled: this.state.schedule !== 'interval' })))),
                react_1.default.createElement("h3", null, "Actions"),
                react_1.default.createElement("p", null, "Reports can be sent by email or posted to a Slack channel. Each report will include the top 10 errors sorted by occurrence."),
                react_1.default.createElement(eui_1.EuiSwitch, { label: "Send email", checked: this.state.actions.email, onChange: () => this.onChangeAction('email') }),
                react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }),
                this.state.actions.email && (react_1.default.createElement(eui_1.EuiFormRow, { label: "Recipients (separated with comma)", compressed: true, helpText: react_1.default.createElement("span", null,
                        "If you have not configured email, please see the",
                        ' ',
                        react_1.default.createElement(eui_1.EuiLink, { target: "_blank", href: xpack_1.XPACK_DOCS.xpackEmails }, "documentation"),
                        ".") },
                    react_1.default.createElement(eui_1.EuiFieldText, { icon: "user", value: this.state.emails, onChange: this.onChangeEmails }))),
                react_1.default.createElement(eui_1.EuiSwitch, { label: "Send Slack notification", checked: this.state.actions.slack, onChange: () => this.onChangeAction('slack') }),
                react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }),
                this.state.actions.slack && (react_1.default.createElement(eui_1.EuiFormRow, { label: "Slack Webhook URL", compressed: true, helpText: react_1.default.createElement("span", null,
                        "To get a Slack webhook, please see the",
                        ' ',
                        react_1.default.createElement(eui_1.EuiLink, { target: "_blank", href: "https://get.slack.help/hc/en-us/articles/115005265063-Incoming-WebHooks-for-Slack" }, "documentation"),
                        ".") },
                    react_1.default.createElement(eui_1.EuiFieldText, { icon: "link", value: this.state.slackUrl, onChange: this.onChangeSlackUrl }))))));
        return (react_1.default.createElement(eui_1.EuiFlyout, { onClose: this.props.onClose, size: "s" },
            react_1.default.createElement(eui_1.EuiFlyoutHeader, null,
                react_1.default.createElement(eui_1.EuiTitle, null,
                    react_1.default.createElement("h2", null, "Enable error reports"))),
            react_1.default.createElement(eui_1.EuiFlyoutBody, null, flyoutBody),
            react_1.default.createElement(eui_1.EuiFlyoutFooter, null,
                react_1.default.createElement(eui_1.EuiButton, { onClick: this.createWatch, fill: true, disabled: !this.state.actions.email && !this.state.actions.slack }, "Create watch"))));
    }
}
exports.WatcherFlyout = WatcherFlyout;
