// x-pack/plugins/beats_management/public/components/tag/config_view/config_form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
// @ts-ignore
const react_1 = require("@kbn/i18n/react");
const formsy_react_1 = tslib_1.__importStar(require("formsy-react"));
const js_yaml_1 = tslib_1.__importDefault(require("js-yaml"));
const lodash_1 = require("lodash");
const react_2 = tslib_1.__importDefault(require("react"));
const inputs_1 = require("../../inputs");
formsy_react_1.addValidationRule('isHosts', (form, values) => {
    // TODO add more validation
    return true;
});
formsy_react_1.addValidationRule('isString', (values, value) => {
    return true;
});
formsy_react_1.addValidationRule('isPeriod', (values, value) => {
    // TODO add more validation
    return true;
});
formsy_react_1.addValidationRule('isPath', (values, value) => {
    // TODO add more validation
    return value && value.length > 0;
});
formsy_react_1.addValidationRule('isPaths', (values, value) => {
    // TODO add more validation
    return true;
});
formsy_react_1.addValidationRule('isYaml', (values, value) => {
    try {
        const stuff = js_yaml_1.default.safeLoad(value || '');
        if (typeof stuff === 'string') {
            return false;
        }
        return true;
    }
    catch (e) {
        return false;
    }
});
class ConfigFormUi extends react_2.default.Component {
    constructor(props) {
        super(props);
        this.form = react_2.default.createRef();
        this.enableButton = () => {
            this.setState({
                canSubmit: true,
            });
            this.props.canSubmit(true);
        };
        this.disableButton = () => {
            this.setState({
                canSubmit: false,
            });
            this.props.canSubmit(false);
        };
        this.submit = () => {
            if (this.form.current && this.props.onSubmit) {
                this.form.current.click();
            }
        };
        this.onValidSubmit = (model) => {
            if (!this.props.onSubmit) {
                return;
            }
            this.props.onSubmit(model);
        };
        this.state = {
            canSubmit: false,
        };
    }
    render() {
        const { intl } = this.props;
        return (react_2.default.createElement("div", null,
            react_2.default.createElement("br", null),
            react_2.default.createElement(formsy_react_1.default, { onValidSubmit: this.onValidSubmit, onValid: this.enableButton, onInvalid: this.disableButton },
                this.props.schema.map(schema => {
                    switch (schema.ui.type) {
                        case 'input':
                            return (react_2.default.createElement(inputs_1.FormsyEuiFieldText, { key: schema.id, id: schema.id, defaultValue: lodash_1.get(this.props, `values.configs[0].${schema.id}`, schema.defaultValue), name: schema.id, disabled: !this.props.onSubmit, helpText: schema.ui.helpText, placeholder: schema.ui.placeholder, label: schema.ui.label, validations: schema.validations, validationError: schema.error, required: schema.required || false }));
                        case 'password':
                            return (react_2.default.createElement(inputs_1.FormsyEuiPasswordText, { key: schema.id, id: schema.id, disabled: !this.props.onSubmit, defaultValue: lodash_1.get(this.props, `values.configs[0].${schema.id}`, schema.defaultValue), name: schema.id, placeholder: schema.ui.placeholder, helpText: schema.ui.helpText, label: schema.ui.label, validations: schema.validations, validationError: schema.error, required: schema.required || false }));
                        case 'multi-input':
                            return (react_2.default.createElement(inputs_1.FormsyEuiMultiFieldText, { key: schema.id, id: schema.id, disabled: !this.props.onSubmit, defaultValue: lodash_1.get(this.props, `values.configs[0].${schema.id}`, schema.defaultValue), name: schema.id, placeholder: schema.ui.placeholder, helpText: schema.ui.helpText, label: schema.ui.label, validations: schema.validations, validationError: schema.error, required: schema.required }));
                        case 'select':
                            return (react_2.default.createElement(inputs_1.FormsyEuiSelect, { key: schema.id, id: schema.id, name: schema.id, disabled: !this.props.onSubmit, defaultValue: lodash_1.get(this.props, `values.configs[0].${schema.id}`, schema.defaultValue), helpText: schema.ui.helpText, label: schema.ui.label, options: [
                                    {
                                        value: '',
                                        text: intl.formatMessage({
                                            id: 'xpack.beatsManagement.tagConfig.selectOptionLabel',
                                            defaultMessage: 'Please Select An Option',
                                        }),
                                    },
                                ].concat(schema.options || []), validations: schema.validations, validationError: schema.error, required: schema.required }));
                        case 'code':
                            return (react_2.default.createElement(inputs_1.FormsyEuiCodeEditor, { key: `${schema.id}-${this.props.id}`, mode: "yaml", disabled: !this.props.onSubmit, id: schema.id, defaultValue: lodash_1.get(this.props, `values.configs[0].${schema.id}`, schema.defaultValue), name: schema.id, helpText: schema.ui.helpText, label: schema.ui.label, options: schema.options ? schema.options : [], validations: schema.validations, validationError: schema.error, required: schema.required }));
                    }
                }),
                this.props.onSubmit && (react_2.default.createElement("button", { type: "submit", style: { display: 'none' }, disabled: !this.state.canSubmit, ref: this.form })))));
    }
}
exports.ConfigForm = react_1.injectI18n(ConfigFormUi, { withRef: true });
