// x-pack/plugins/beats_management/server/rest_api/tags/get.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("x-pack/plugins/beats_management/common/constants");
const error_wrappers_1 = require("../../utils/error_wrappers");
exports.createGetTagsWithIdsRoute = (libs) => ({
    method: 'GET',
    path: '/api/beats/tags/{tagIds}',
    requiredRoles: ['beats_admin'],
    licenseRequired: constants_1.REQUIRED_LICENSES,
    handler: async (request) => {
        const tagIdString = request.params.tagIds;
        const tagIds = tagIdString.split(',').filter((id) => id.length > 0);
        let tags;
        try {
            tags = await libs.tags.getTagsWithIds(request.user, tagIds);
        }
        catch (err) {
            return error_wrappers_1.wrapEsError(err);
        }
        return tags;
    },
});
