// x-pack/plugins/apm/public/components/app/ErrorGroupDetails/DetailView/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const lodash_1 = require("lodash");
const react_1 = tslib_1.__importDefault(require("react"));
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
const constants_1 = require("x-pack/plugins/apm/common/constants");
const constants_2 = require("../../../../../common/constants");
const constants_3 = require("../../../../constants");
const variables_1 = require("../../../../style/variables");
const url_1 = require("../../../../utils/url");
const url_2 = require("../../../../utils/url");
const DiscoverErrorButton_1 = require("../../../shared/DiscoverButtons/DiscoverErrorButton");
const PropertiesTable_1 = require("../../../shared/PropertiesTable");
const Stacktrace_1 = require("../../../shared/Stacktrace");
const StickyProperties_1 = require("../../../shared/StickyProperties");
const Container = styled_components_1.default.div `
  position: relative;
  border: 1px solid ${variables_1.colors.gray4};
  border-radius: ${variables_1.borderRadius};
  margin-top: ${variables_1.px(variables_1.units.plus)};
`;
const HeaderContainer = styled_components_1.default.div `
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  padding: ${variables_1.px(variables_1.units.plus)} ${variables_1.px(variables_1.units.plus)} 0;
  margin-bottom: ${variables_1.px(variables_1.unit)};
`;
const PaddedContainer = styled_components_1.default.div `
  padding: ${variables_1.px(variables_1.units.plus)} ${variables_1.px(variables_1.units.plus)} 0;
`;
const EXC_STACKTRACE_TAB = 'exception_stacktrace';
const LOG_STACKTRACE_TAB = 'log_stacktrace';
function DetailView({ errorGroup, urlParams, location }) {
    if (errorGroup.status !== constants_3.STATUS.SUCCESS) {
        return null;
    }
    const { transaction, error, occurrencesCount } = errorGroup.data;
    if (!error) {
        return null;
    }
    const transactionLink = getTransactionLink(error, transaction);
    const stickyProperties = [
        {
            fieldName: '@timestamp',
            label: 'Timestamp',
            val: error['@timestamp'],
            width: '50%'
        },
        {
            fieldName: constants_2.REQUEST_URL_FULL,
            label: 'URL',
            val: lodash_1.get(error, constants_2.REQUEST_URL_FULL, 'N/A'),
            truncated: true,
            width: '50%'
        },
        {
            fieldName: constants_2.REQUEST_METHOD,
            label: 'Request method',
            val: lodash_1.get(error, constants_2.REQUEST_METHOD, 'N/A'),
            width: '25%'
        },
        {
            fieldName: constants_2.ERROR_EXC_HANDLED,
            label: 'Handled',
            val: lodash_1.get(error, constants_2.ERROR_EXC_HANDLED, 'N/A'),
            width: '25%'
        },
        {
            fieldName: constants_2.TRANSACTION_ID,
            label: 'Transaction sample ID',
            val: transactionLink || 'N/A',
            width: '25%'
        },
        {
            fieldName: constants_2.USER_ID,
            label: 'User ID',
            val: lodash_1.get(error, constants_2.USER_ID, 'N/A'),
            width: '25%'
        }
    ];
    const tabs = getTabs(error);
    const currentTab = getCurrentTab(tabs, urlParams.detailTab);
    return (react_1.default.createElement(Container, null,
        react_1.default.createElement(HeaderContainer, null,
            react_1.default.createElement(eui_1.EuiTitle, { size: "s" },
                react_1.default.createElement("h3", null, "Error occurrence")),
            react_1.default.createElement(DiscoverErrorButton_1.DiscoverErrorButton, { error: error, kuery: urlParams.kuery },
                react_1.default.createElement(eui_1.EuiButtonEmpty, { iconType: "discoverApp" }, `View ${occurrencesCount} occurrences in Discover`))),
        react_1.default.createElement(PaddedContainer, null,
            react_1.default.createElement(StickyProperties_1.StickyProperties, { stickyProperties: stickyProperties })),
        react_1.default.createElement(eui_1.EuiSpacer, null),
        react_1.default.createElement(eui_1.EuiTabs, null, tabs.map(key => {
            return (react_1.default.createElement(eui_1.EuiTab, { onClick: () => {
                    url_1.history.replace({
                        ...location,
                        search: url_1.fromQuery({
                            ...url_1.toQuery(location.search),
                            detailTab: key
                        })
                    });
                }, isSelected: currentTab === key, key: key }, lodash_1.capitalize(key.replace('_', ' '))));
        })),
        react_1.default.createElement(PaddedContainer, null,
            react_1.default.createElement(TabContent, { error: error, currentTab: currentTab }))));
}
exports.DetailView = DetailView;
function getTransactionLink(error, transaction) {
    if (!transaction || !lodash_1.get(error, 'transaction.sampled')) {
        return;
    }
    const path = `/${transaction.context.service.name}/transactions/${url_2.legacyEncodeURIComponent(transaction.transaction.type)}/${url_2.legacyEncodeURIComponent(transaction.transaction.name)}`;
    return (react_1.default.createElement(url_2.KibanaLink, { pathname: '/app/apm', hash: path, query: {
            transactionId: transaction.transaction.id,
            traceId: lodash_1.get(transaction, constants_2.TRACE_ID)
        } }, transaction.transaction.id));
}
function TabContent({ error, currentTab }) {
    const codeLanguage = error.context.service.name;
    const agentName = error.context.service.agent.name;
    const excStackframes = lodash_1.get(error, constants_1.ERROR_EXC_STACKTRACE);
    const logStackframes = lodash_1.get(error, constants_1.ERROR_LOG_STACKTRACE);
    switch (currentTab) {
        case LOG_STACKTRACE_TAB:
        case undefined:
            return (react_1.default.createElement(Stacktrace_1.Stacktrace, { stackframes: logStackframes, codeLanguage: codeLanguage }));
        case EXC_STACKTRACE_TAB:
            return (react_1.default.createElement(Stacktrace_1.Stacktrace, { stackframes: excStackframes, codeLanguage: codeLanguage }));
        default:
            const propData = error.context[currentTab];
            return (react_1.default.createElement(PropertiesTable_1.PropertiesTable, { propData: propData, propKey: currentTab, agentName: agentName }));
    }
}
exports.TabContent = TabContent;
// Ensure the selected tab exists or use the first
function getCurrentTab(tabs = [], selectedTab) {
    return tabs.includes(selectedTab) ? selectedTab : tabs[0];
}
exports.getCurrentTab = getCurrentTab;
function getTabs(error) {
    const hasLogStacktrace = lodash_1.get(error, constants_1.ERROR_LOG_STACKTRACE, []).length > 0;
    const contextKeys = Object.keys(error.context);
    return [
        ...(hasLogStacktrace ? [LOG_STACKTRACE_TAB] : []),
        EXC_STACKTRACE_TAB,
        ...PropertiesTable_1.getPropertyTabNames(contextKeys)
    ];
}
exports.getTabs = getTabs;
