// x-pack/plugins/apm/server/lib/metrics/get_cpu_chart_data/fetcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("x-pack/plugins/apm/common/constants");
const metricsFetcher_1 = require("../metricsFetcher");
async function fetch(args) {
    return metricsFetcher_1.fetchMetrics({
        ...args,
        timeseriesBucketAggregations: {
            systemCPUAverage: { avg: { field: constants_1.METRIC_SYSTEM_CPU_PERCENT } },
            systemCPUMax: { max: { field: constants_1.METRIC_SYSTEM_CPU_PERCENT } },
            processCPUAverage: { avg: { field: constants_1.METRIC_PROCESS_CPU_PERCENT } },
            processCPUMax: { max: { field: constants_1.METRIC_PROCESS_CPU_PERCENT } }
        },
        totalAggregations: {
            systemCPUAverage: { avg: { field: constants_1.METRIC_SYSTEM_CPU_PERCENT } },
            systemCPUMax: { max: { field: constants_1.METRIC_SYSTEM_CPU_PERCENT } },
            processCPUAverage: { avg: { field: constants_1.METRIC_PROCESS_CPU_PERCENT } },
            processCPUMax: { max: { field: constants_1.METRIC_PROCESS_CPU_PERCENT } }
        }
    });
}
exports.fetch = fetch;
